oak_restricted_kernel/src/payload.rs

```diff
@@ -128,33 +133,112 @@ impl Application {
             self.load_segment(phdr).unwrap();
         }
 
-        // Set up the userspace stack at the end of the lower half of the virtual address space.
-        // Well... almost. It's one page lower than the very end, as otherwise the initial stack
-        // pointer would need to be a noncanonical address, so let's avoid that whole mess
-        // by moving down a bit.
-        let rsp = VirtAddr::new(0x8000_0000_0000 - Size2MiB::SIZE);
         mmap(
-            Some(rsp - Size2MiB::SIZE),
+            Some(VirtAddr::new(APPLICATION_STACK_VIRT_ADDR) - Size2MiB::SIZE),
             Size2MiB::SIZE as usize,
             MmapProtection::PROT_READ | MmapProtection::PROT_WRITE,
             MmapFlags::MAP_ANONYMOUS | MmapFlags::MAP_PRIVATE | MmapFlags::MAP_FIXED,
         )
         .expect("failed to allocate memory for user stack");
 
-        log::info!("Running application");
+        self.entry()
+    }
+}
+
+pub struct Process {
+    pml4: x86_64::structures::paging::PageTable,
+    entry: VirtAddr,
+}
+
+impl Process {
+    /// Creates a process from the application, without executing it.
+    ///
+    /// # Safety
+    ///
+    /// The application must be built from a valid ELF file representing an Oak Restricted
+    /// Application.
+    pub unsafe fn from_application(application: &Application) -> Result<Self, anyhow::Error> {
+        let pml4 = crate::BASE_L4_PAGE_TABLE
+            .get()
+            .context("base l4 table should be set")?
+            .clone();
+        // Load the process's page table, so the application can be loaded into its memory. Hold
+        // onto the previous PT, so we can revert to it once the application has been mapped
+        // into the process pt.
```

Comment:
Instead of loading in the process's pt and then reverting at the end of the function, we could map the memory into the pt without switching to it, by passing it as an arg to mmap. Imo nicer, but out of scope for now.